src/dsp_tools/utils/xmlupload/resource_create_client.py

```diff
@@ -0,0 +1,379 @@
+        case _:
+            raise UserError(f"Unknown file ending: {file_ending} for file {local_file}")
+    file_value = {
+        "@type": f"knora-api:{value_type}",
+        "knora-api:fileValueHasFilename": bitstream_info.internal_file_name,
+    }
+    if bitstream_info.permissions:
+        file_value["knora-api:hasPermissions"] = str(bitstream_info.permissions)
+    return {prop: file_value}
+
+
+def _make_boolean_value(value: XMLValue) -> dict[str, Any]:
+    return {
+        "@type": "knora-api:BooleanValue",
+        "knora-api:booleanValueAsBoolean": bool(value.value),
+    }
+
+
+def _make_color_value(value: XMLValue) -> dict[str, Any]:
+    return {
+        "@type": "knora-api:ColorValue",
+        "knora-api:colorValueAsColor": value.value,
+    }
+
+
+def _make_date_value(value: XMLValue) -> dict[str, Any]:
+    entire_pattern = r"""
+    ^
+    (?:(GREGORIAN|JULIAN|ISLAMIC):)?        # optional calendar
+    (?:(CE|BCE|BC|AD):)?                    # optional era
+    (\d{4}(?:-\d{1,2})?(?:-\d{1,2})?)       # date
+    (?::(CE|BCE|BC|AD))?                    # optional era
+    (?::(\d{4}(?:-\d{1,2})?(?:-\d{1,2})?))? # optional date
+    $
+"""
+    assert isinstance(value.value, str)
+    date_match = regex.search(entire_pattern, value.value, flags=regex.VERBOSE)
+    assert date_match
+    date_groups = date_match.groups()
+    match date_groups:
+        case (str() | None, str() | None, str(), str() | None, str() | None):
+            # start_era cannot be None, and this provides a useful type hint
+            pass
+        case ("ISLAMIC", era, _, end_era, _) if era or end_era:
+            raise UserError(f"ISLAMIC calendar does not support eras: {value.value}")
+        case _:
+            raise UserError(f"Could not parse date: {value.value}")
+    res: dict[str, Any] = {"@type": "knora-api:DateValue"}
+    calendar, start_era, start_date, end_era, end_date = date_groups
+    if not calendar:
+        calendar = "GREGORIAN"
+    if not end_date:
+        end_date = start_date
+    if calendar != "ISLAMIC":
+        if not start_era:
+            start_era = "CE"
+        if end_date and not end_era:
+            end_era = "CE"
+
+    year, month, day = _parse_date(start_date)
+    res["knora-api:dateValueHasStartYear"] = int(year)
+    if month:
+        res["knora-api:dateValueHasStartMonth"] = int(month)
+    if day:
+        res["knora-api:dateValueHasStartDay"] = int(day)
+    if end_era:
+        res["knora-api:dateValueHasEndEra"] = end_era
+
+    if calendar:
+        res["knora-api:dateValueHasCalendar"] = calendar
+    if start_era:
+        res["knora-api:dateValueHasStartEra"] = start_era
+    if end_date:
+        year, month, day = _parse_date(end_date)
+        res["knora-api:dateValueHasEndYear"] = int(year)
+        if month:
+            res["knora-api:dateValueHasEndMonth"] = int(month)
+        if day:
+            res["knora-api:dateValueHasEndDay"] = int(day)
+    return res
```

Comment:
I'm not a fan of how this turned out, probably some of this logic should be factored out to keep it manageable. What do you think?